### PR TITLE
[Feat][TPU Offload] KV cache offload to local cpu buffer

### PR DESCRIPTION
**Pull request by juncgu-google**
# Description

TL;DR, add the feature of offloading KV cache to host cpu buffer (similar to the [native CPU offloading in vLLM](https://github.com/vllm-project/vllm/issues/19854)).

This PR allows offloading computed KV cache (at the granularity of block / page) (of prompt tokens or even including the generated tokens) to the host CPU buffer and bringing them back to TPU HBM when there are cache hits, to avoid re-compute. 

## implementation
Following the [general kv connector interfaces](https://github.com/vllm-project/vllm/blob/main/vllm/distributed/kv_transfer/kv_connector/v1/base.py) and the logic of [native CPU offloading in vLLM](https://github.com/vllm-project/vllm/issues/19854), it introduces a  [TPUOffloadConnector](https://github.com/vllm-project/tpu-inference/blob/87db579613f1bb42e4529487645f510538416dde/tpu_inference/distributed/offload/tpu_offload_connector.py#L336), which is the central component of managing and executing the offloading logic. Within the `TPUOffloadConnector`, there are:
1. [TPUOffloadConnectorScheduler](https://github.com/vllm-project/tpu-inference/blob/87db579613f1bb42e4529487645f510538416dde/tpu_inference/distributed/offload/tpu_offload_connector.py#L470): colocates with the vLLM scheduler, makes KV cache swap decisions, and keeps track of running requests and their in-flight swap operations.   
   - [OffloadManager](https://github.com/vllm-project/tpu-inference/blob/87db579613f1bb42e4529487645f510538416dde/tpu_inference/distributed/offload/tpu_offload_connector.py#L479): manages the logical kv cache space on host CPU buffer, and applys LRU eviction policy.
   - [StagingBufferManager](https://github.com/vllm-project/tpu-inference/blob/87db579613f1bb42e4529487645f510538416dde/tpu_inference/distributed/offload/tpu_offload_connector.py#L514): manages a logical staging buffer in the TPU's HBM, which is used as a temporary buffer (in HBM) for data being transferred between the TPU and the CPU. Without the control of using temporary buffer in HBM can easily lead to TPU OOM.  
 2. [TPUOffloadConnectorWorker](https://github.com/vllm-project/tpu-inference/blob/87db579613f1bb42e4529487645f510538416dde/tpu_inference/distributed/offload/tpu_offload_connector.py#L1150): executes the swap decisions (embedded in the SchedulerOutput) from `TPUOffloadConnectorScheduler`.     
    - [LocalCPUBackend](https://github.com/vllm-project/tpu-inference/blob/87db579613f1bb42e4529487645f510538416dde/tpu_inference/distributed/offload/tpu_offload_connector.py#L1172): a simple key-value map in host CPU memory for hosting the offloaded KV cache blocks / pages.
 
The `TPU-CPU` swap operations are the core of this PR. We provide two approaches to move KV cache data:
1. using  [jax.device_put](https://github.com/vllm-project/tpu-inference/blob/87db579613f1bb42e4529487645f510538416dde/tpu_inference/distributed/offload/utils.py#L125)
2. using [pallas dma kernel](https://github.com/vllm-project/tpu-inference/blob/87db579613f1bb42e4529487645f510538416dde/tpu_inference/distributed/offload/utils.py#L148)

## usage
This feature can be used by setting the `kv_tranfer_config` in vLLM engine:
`--kv-transfer-config '{"kv_connector":"TPUOffloadConnector","kv_connector_module_path":"tpu_inference.distributed.offload.tpu_offload_connector","kv_role":"kv_both"}'`
example: [examples/offload/gke/benchmarks/deploy-cpu-offload.yaml](https://github.com/vllm-project/tpu-inference/blob/cpu-offloading/dev-merge/examples/offload/gke/benchmarks/deploy-cpu-offload.yaml#L24)

And, it can be configured through the following [environment variables](https://github.com/vllm-project/tpu-inference/blob/87db579613f1bb42e4529487645f510538416dde/tpu_inference/envs.py#L27-L31) (we will move them into the `KVTransferConfig.kv_connector_extra_config`):
1. `TPU_OFFLOAD_SKIP_JAX_PRECOMPILE`: skipping pre-compiling the swap functions, default=0. We would suggest to turn on pre-compile. All swap operations are applied at block-granularity; when swap pre-compile is turned on, we will break a request of swap into multiple swap-operations following the predefined bucket size list (1 block, 2 blocks, 4 blocks, 8 blocks, 16 blocks) to avoid re-compile (thanks to @saikat-royc).
2. `TPU_OFFLOAD_SWAP_OP_TYPE`: `jax` (default), or `pallas`.
3. `TPU_OFFLOAD_NUM_CPU_CHUNKS`: host CPU buffer capacity in terms of number of chunks (equivalent to kv cache blocks / pages), default=1024.
4. `TPU_OFFLOAD_NUM_STAGING_BLOCKS`: the size of the staging buffer in the TPU HBM, default=128.
5. `TPU_OFFLOAD_DECODE_SAVE`: save the KV cache of generated (decode) tokens, default=False.

# Tests

`pytest -s -v tests/distributed/offload/`
`pytest -s -v tests/kernels/host_dma_test.py`

# Checklist

Before submitting this PR, please make sure:
- I have performed a self-review of my code.
- I have necessary comments in my code, particularly in hard-to-understand areas.
- I have made or will make corresponding changes to any relevant documentation.

### CODE REVIEW

File: examples/offline_inference_kv_cache_verification.py [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+"""
+
+import copy
+import os
+import time
+from typing import List, Tuple
+
+import vllm.envs as envs
+from vllm import LLM, EngineArgs, SamplingParams
+from vllm.utils import FlexibleArgumentParser
```
Comment (dannawang0221) [inline examples/offline_inference_kv_cache_verification.py]: This needs to be updated to `from vllm.utils.argparse_utils import FlexibleArgumentParser` The docker image now is pulling from the vllm main branch, and lib got moved to another place

File: examples/offline_inference_kv_cache.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
+# SPDX-License-Identifier: Apache-2.0
+
+import os
+import time
+
+import vllm.envs as envs
+from vllm import LLM, EngineArgs
+from vllm.utils import FlexibleArgumentParser
```
Comment (dannawang0221) [inline examples/offline_inference_kv_cache.py]: Same here. This needs to be updated to `from vllm.utils.argparse_utils import FlexibleArgumentParser`

File: examples/offload/gke/pod_tpu_host_offload_unit_tests.yaml [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    imagePullPolicy: Always
+    command:
+    - /bin/bash
+    - -c
+    - "pytest -sv tests/distributed/offload/tpu_offload_cpu_backend_test.py"
+    - "pytest -sv tests/distributed/offload/tpu_offload_connector_worker_test.py"
+    - "pytest -sv tests/distributed/offload/tpu_offload_connector_scheduler_test.py"
+    - "pytest -sv tests/distributed/offload/tpu_offload_utils_test.py"
+    - "pytest -sv tests/distributed/offload/tpu_offload_manager_test.py"
+    - "pytest -sv tests/distributed/offload/tpu_offload_accuracy_test.py"
```
Comment (dannawang0221) [inline examples/offload/gke/pod_tpu_host_offload_unit_tests.yaml]: Could you replace line 20 - line 25 with `- "pytest -sv tests/distributed/offload/"`? The current code only run the test on line 20

File: tests/offload/tpu_offload_connector_worker_test.py [hunk 4]
```diff
@@ -0,0 +1,533 @@
...
+            2,
+            self.head_size,
+        )
+        self.cache_dtype = jnp.bfloat16
+        partition_spec = PartitionSpec(None, None, "model")
+        self.device_sharding = NamedSharding(self.mesh, partition_spec)
+
+    def tearDown(self):
+        super().tearDown()
+        cc.reset_cache()
```
Comment (dannawang0221) [inline tests/offload/tpu_offload_connector_worker_test.py:102]: Could you add `jax.clear_caches()` after `cc.reset_cache()`? The test crash with error
```
INFO 12-01 05:19:26 ... Compiling for 16 blocks...
F1201 05:19:27.965941 951 alignment_util.cc:17] Check failed: err == 0 (12 vs. 0) posix_memalign failed: Cannot allocate memory
...
F1201 05:19:27.966116 951 low_level_alloc.cc:619] RAW: mmap error: 12
```

**Comment by kyuyeunk** [comment]
this is really really big pr & it's difficult to review.

are you planning to split this PR into multiple smaller ones to make it easier for reviewers?

File: tests/offload/tpu_offload_connector_worker_test.py [hunk 5]
```diff
@@ -0,0 +1,565 @@
...
+        jax.clear_caches()
```
Comment (dannawang0221) [inline tests/offload/tpu_offload_connector_worker_test.py:103]: I hit the same error again, seems like the previous change didn't completely fix the issue ```INFO 12-01 05:19:26 [tpu_offload_connector.py:1311]     - Calling jitted_insert_kv_cache_slices with paged_kv_for_compilation len: 80, first_element_shape: (128, 256, 8, 2, 128), chunked_dummy_kv_tpu len: 80
INFO 12-01 05:19:26 [tpu_offload_connector.py:1287]   - Compiling for 16 blocks...
F1201 05:19:27.965941     951 alignment_util.cc:17] Check failed: err == 0 (12 vs. 0) posix_memalign failed: Cannot allocate memory
F1201 05:19:27.965968     949 alignment_util.cc:17] Check failed: err == 0 (12 vs. 0) posix_memalign failed: Cannot allocate memory
F1201 05:19:27.965991     945 alignment_util.cc:17] Check failed: err == 0 (12 vs. 0) posix_memalign failed: Cannot allocate memory
F1201 05:19:27.966026     950 alignment_util.cc:17] Check failed: err == 0 (12 vs. 0) posix_memalign failed: Cannot allocate memory
F1201 05:19:27.966114     946 alignment_util.cc:17] Check failed: err == 0 (12 vs. 0) posix_memalign failed: Cannot allocate memory
F1201 05:19:27.966116     951 low_level_alloc.cc:619] RAW: mmap error: 12```

There are two other suggestions:
1. We can try more aggressive memory release 
```        
        Destroy references explicitly
        if hasattr(self, 'connector'):
              del self.connector
        
        # Force JAX to release memory
        cc.reset_cache()
         jax.clear_caches()
        
        # Force Python GC
        gc.collect()
```
2. Can we change ` self.num_layers ` to a smaller number since this is a unit test

Comment (juncgu-google) [reply tests/offload/tpu_offload_connector_worker_test.py:103]: resolved by e8bc4ac612a9bda8d7ac65f46f91a166c4c5ef7d

**Comment by juncgu-google** [comment]
> this is really really big pr & it's difficult to review.
> 
> are you planning to split this PR into multiple smaller ones to make it easier for reviewers?

That's true. The core implementation is in the distributed/offload folder (~2400 lines), others are for tests. But this feature is modularized and makes negligible changes to the core of tpu-inference. 
I will discuss with all of the reviewers to see how to re-organize the PR.

**Review by kyuyeunk — CHANGES_REQUESTED** [review]
Can you move the implementation outside of the folder "distributed" since this feature can be used for non-distributed case as well?

**Comment by juncgu-google** [comment]
> Can you move the implementation outside of the folder "distributed" since this feature can be used for non-distributed case as well?

thanks for the comment. done.

File: tpu_inference/offload/tpu_offload_connector.py [hunk 6]
```diff
@@ -1219,8 +1219,9 @@ def __init__(self, vllm_config: VllmConfig,
 
         self.cpu_chunk_size = self.block_size
         # Thread pool for asynchronous TPU->CPU copies
+        self.num_save_threads = int(os.getenv("TPU_OFFLOAD_SAVE_THREADS", "1"))
```
Comment (juncgu-google) [inline tpu_inference/offload/tpu_offload_connector.py]: please refer to [TPU_OFFLOAD_SWAP_OP_TYPE](https://github.com/vllm-project/tpu-inference/blob/b659166c375cc52a9e2b7d4437aeab0c2d45f6d5/tpu_inference/envs.py#L28) and update the way of reading env var.

Comment (dannawang0221) [reply tpu_inference/offload/tpu_offload_connector.py]: Updated
